packages/replayer/src/replay.utils.ts

```diff
@@ -130,31 +130,84 @@ const installVirtualEventLoop = (page: Page, sessionStartTime: DateTime) => {
+    );
+    return DateTime.now().toUTC();
+  }
+
+  // event.timeStamp differs from event.timeStampRaw in that (a) it is relative to the start of the session,
+  // (b) it has been adjusted to subtract the time between constructing the recorder and calling.start().
+  //
+  // By subtracting event.timeStamp from event.timeStampRaw we get the performance.timeOrigin plus the
+  // the time between constructing the recorder and calling.start(), in other words: the time at which recording
+  // started.
+  //
+  // Note that performance.timeOrigin uses monotonic clock time, which can differ by multiple hours
+  // from the time returned by Date.now(). We may in future want to record a dedicated
+  // session start time to ensure we fully accuractly recreate the original values returned by the
+  // Date.now() calls.
+  const minUserEventTimestamp =
+    userEventEventLog[0].timeStampRaw - userEventEventLog[0].timeStamp;
+  return DateTime.fromMillis(minUserEventTimestamp).toUTC();
+};
 
-export const getRrwebRecordingDuration: (
+export const getSessionDuration: (
   sessionData: SessionData
 ) => Duration | null = (sessionData) => {
-  const [minRrwebTimestamp, maxRrwebTimestamp] =
-    getMinMaxRrwebTimestamps(sessionData);
-  const rrwebRecordingDuration = maxRrwebTimestamp.diff(minRrwebTimestamp);
-  return rrwebRecordingDuration.isValid ? rrwebRecordingDuration : null;
+  const rrWebTimeRange = getMinMaxRrwebTimestamps(sessionData);
+  const rrWebDuration =
+    rrWebTimeRange != null ? rrWebTimeRange.max.diff(rrWebTimeRange.min) : null;
+
+  if (rrWebDuration != null && rrWebDuration.isValid) {
+    return rrWebDuration;
+  }
+
+  const userEventEventLog = sessionData.userEvents?.event_log ?? [];
+
+  // The replayer uses the event timeStamps to work out how long to wait before replaying each event.
+  // We want the amount of time we pad at the end to be consistent with this, and so we use the same
+  // timestamps.
+  //
+  // event.timeStamp differs from event.timeStampRaw in that (a) it is relative to the start of the session,
+  // (b) it has been adjusted to subtract the time between constructing the recorder and calling.start().
+  const maxUserEventTimestamp =
+    userEventEventLog[userEventEventLog.length - 1]?.timeStamp;
+
+  if (maxUserEventTimestamp == null) {
+    return null; // Cannot calculate the duration if no user events or rrweb events
+  }
+
+  return Duration.fromMillis(maxUserEventTimestamp);
 };
 
-const getMinMaxRrwebTimestamps: (
+const getMinMaxRrwebTimestamps = (
   sessionData: SessionData
-) => [DateTime, DateTime] = (sessionData) => {
+): { min: DateTime; max: DateTime } | null => {
   const rrwebTimestamps = (sessionData.rrwebEvents as { timestamp?: number }[])
     .map((event) => event.timestamp || NaN)
     .filter((ts) => !isNaN(ts));
+
+  if (rrwebTimestamps.length === 0) {
+    return null;
+  }
```

Comment:
`luxon` correctly propagates invalid timestamp computations, so this is not strictly needed

Reply:
It's correctly that the code could be made to work without this. You'd get DateTime.fromMillis of Infinity which would return invalid timestamps, which you could then check in getSessionDuration. And that's how I initially implemented it, however I refactored to this to ensure it's explicit -- the reader can understand that we intentionally handle the empty case in a particular way, without having to (a) know the behaviour of Math.max on an empty set, (b) the behaviour of luxon on infinite values, (c) that we wrote the code intentionally taking advantage of these behaviours, rather than accidentally, to handle a special case of the rrwebTimestamps array being empty.

Reply:
General rule here is that code that is quicker and easier to understand the programmers intent and the intended flows of the code / different cases is better even if the code has more lines/characters.

Reply:
:+1: